src/Lecture2.hs

```diff
@@ -155,12 +176,87 @@ You're free to define any helper functions.
 
 -- some help in the beginning ;)
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
+    { knightHealth     :: Int
+    , knightAttack     :: Int
+    , knightEndurance  :: Int
+    , knightExperience :: Int
+    , knightGold       :: Int
+    , knightSword      :: Maybe Sword
+    , knightShield     :: Maybe Shield
+    , knightAlive      :: Bool
+    }
+    deriving (Show)
+
+data Color
+    = Red
+    | Black
+    | Green
+    deriving (Show, Eq, Ord)
+
+{-
+type RedDragonChest = Chest Shield
+type BlackDragonChest = Chest Sword
+type GreenDragonChest = Chest
+-}
+
+data Sword = Sword
+    { swordAttack :: Int
     }
+    deriving (Show)
 
-dragonFight = error "TODO"
+data Shield = Shield
+    { shieldDefense :: Int
+    }   
+    deriving (Show)
+
+data Chest a = Chest
+    { chestGold :: Word
+    , chestTreasure :: a
+    }
+    deriving (Show)
+
+data Dragon = Dragon
+    { dragonColor :: Color
+    , dragonExperience :: Int
+    , dragonHealth :: Int
+    , dragonFirePower :: Int
+    }
+    deriving (Show)
+
+hatchDragon :: Color -> Dragon
+hatchDragon Red = Dragon Red 100 100 10
+hatchDragon Black = Dragon Black 150 150 15
+hatchDragon Green = Dragon Green 250 250 25
+
+{-
+reward :: Dragon -> Chest a
+reward d
+    | dragonColor d == Red = Chest 100 (Shield 1)
+    | dragonColor d == Black = Chest 150 (Sword 1)
+    | dragonColor d == Green = Chest 250
```

Comment:
You probably noticed a difficulty with this function is that `Chest` needs to have two values. So you can't pass only gold here in the last case.

To solve this problem and to model the task in the most precise way, you can define the following types:

```haskell
data Dragon = Dragon
    { dragonExperience :: Int
    , dragonHealth :: Int
    , dragonFirePower :: Int
    }

data DragonType a
    = Green Dragon Gold
    | Red Dragon (Chest a)
    | Black Dragon (Chest a)
```

This way you can easily extract the reward from `Dragon` 🙂

Reply:
Yeah, that was the hard part for me!